### ISSUE TITLE
tf.image.extract_patches fails to JIT-compile inside keras model

### ISSUE THREAD

**Issue by shkarupa-alex** [tensorflow/tensorflow#59058]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Have you reproduced the bug with TF nightly?

Yes

### Source

binary

### Tensorflow Version

2.11.0

### Custom Code

Yes

### OS Platform and Distribution

Ubuntu 20.04, Mac OS X 12.6

### Python version

3.8

### Current Behaviour?

```shell
When trying to set jit_compile=True at model with Dense (Conv2D) layer followed by tf.image.extract_patches an error occured.
```

### Standalone code to reproduce the issue

```shell
https://colab.research.google.com/drive/1-lXiMNp8NSe-c47-NAruypplraDId3O3?usp=sharing
```

### Relevant log output

```shell
Detected unsupported operations when trying to compile graph __inference_run_step_742[] on XLA_GPU_JIT: SparseTensorDenseMatMul (No registered 'SparseTensorDenseMatMul' OpKernel for XLA_GPU_JIT devices compatible with node {{node gradient_tape/model/temp/SparseTensorDenseMatMul/SparseTensorDenseMatMul}}){{node gradient_tape/model/temp/SparseTensorDenseMatMul/SparseTensorDenseMatMul}}
```
```
</details>

**Comment by mohantym** [comment]
Hi @shkarupa-alex !
I was able to replicate this bug with tf.function style and work-around is to use auto-clustering. 

@SuryanarayanaY !
Could you look at this issue. Attached gist in[ 2.9](https://colab.research.google.com/gist/mohantym/47012c653026b233d33044b56c30d8b4/extract_patches_issue.ipynb), [2.10](https://colab.research.google.com/gist/mohantym/fd5d8b7bfe5aa10001b25b0cc6a46a5a/extract_patches_issue.ipynb#scrollTo=6QgMUKYphDm9), [2.11](https://colab.research.google.com/gist/mohantym/fd5d8b7bfe5aa10001b25b0cc6a46a5a/extract_patches_issue.ipynb#scrollTo=8GX2V_LKhNrd) and [nightly ](https://colab.research.google.com/gist/mohantym/1bd22aad7cdcaabbb15ec4fadc753980/extract_patches_issue.ipynb#scrollTo=6QgMUKYphDm9)for reference.

Thank you!

**Comment by rajivhockey97** [comment]
The function tf.image.extract_patches is not supported on cloud TPU. Check out this link for more information https://cloud.google.com/tpu/docs/tensorflow-ops
@mohantym In you auto-clustering code, the code fails if we try to set jit_compile=True in model.compile function. Can you elaborate more about the auto-clustering code?

**Comment by SuryanarayanaY** [comment]
Hi @rajivhockey97 ,
XLA have some limitations.At present all the Ops are not compilable with `jit_compile=True`.There are also some known errors as documented [here](https://www.tensorflow.org/xla/known_issues). The error might be probably due to either the Op is not compatible with Jit_compile or the shapes might be dynamic which is not supported by XLA.

If you choose auto_clustering then it will automatically chooses from CPU to GPU based on implementation which is not possible with explicit setting in model.compile(jit_compile=True) which forces to use only `jit_compile` and hence throwing the error. Please refer to autoclustering tutorial [here](https://www.tensorflow.org/xla/tutorials/autoclustering_xla). However `autoclustering` too have limitations at present like below as per [source](https://www.tensorflow.org/xla#auto-clustering).

> Note: In TF2, only the code inside `tf.function` will be clustered.

> Note: Auto-clustering support on CPU and on multi-GPU environments is experimental.

It seems the shape of the Op `tf.image.extract_patches` is not inferrable at the time of compilation i.e. the output shape of the Op output can't be inferred without actual computation which might happen only during `model.fit()` in this case.Could you cross check and confirm?

**Comment by rajivhockey97** [comment]
@SuryanarayanaY Thanks for the explanation. Now it makes sense as to why the error shows up when jit_compile is set to true. However i am not sure how to identify whether the shape of the Op tf.image.extract_patches is inferrable at the time of the compilation or not.

**Comment by SuryanarayanaY** [comment]
Hi @rajivhockey97 ,

If the shape is not inferrable without some computation that may happens during `model.fit()` or say during training step then it is not available for compilation.
In our case we can check it by `model.summary()` before `model.compile()`.If we don't get the proper shape(of course batch size might be neglected here since we pass it during training only) may be we can't use `jit_compile` there. Please refer the demo [gist](https://colab.research.google.com/gist/SuryanarayanaY/9eccad58662e2628a7cf588d714e75da/59058_r2.ipynb) for our case.Hope it helps for you and comeback if still needs clarity.

Thankyou!

**Comment by shkarupa-alex** [comment]
> Hi @rajivhockey97 ,
> 
> If the shape is not inferrable without some computation that may happens during `model.fit()` or say during training step then it is not available for compilation. In our case we can check it by `model.summary()` before `model.compile()`.If we don't get the proper shape(of course batch size might be neglected here since we pass it during training only) may be we can't use `jit_compile` there. Please refer the demo [gist](https://colab.research.google.com/gist/SuryanarayanaY/9eccad58662e2628a7cf588d714e75da/59058_r2.ipynb) for our case.Hope it helps for you and comeback if still needs clarity.
> 
> Thankyou!

Even if we set constant shape for input layer in this gist (and all shapes in model summary become defined), it still fails with jit_compile=True

**Comment by shkarupa-alex** [comment]
As far as i understand, this issue along with https://github.com/tensorflow/tensorflow/issues/59061 come from gradient implementation here https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/array_grad.py#L1021

**Comment by SuryanarayanaY** [comment]
@shkarupa-alex ,
Even if the input shapes are known but if the function output shapes depends upon computation that happening inside the function then jit_compile=True may not work. For example let take the sample code of yours below:
```
    def call(self, inputs):
        outputs = self.d1(inputs)
        outputs = tf.image.extract_patches(outputs, [1, 16, 16, 1], [1, 8, 8, 1], [1] * 4, padding='SAME')
        outputs = self.d2(outputs)
        
        return outputs
```
The shape of the variable `outputs` might depends upon the parameters `([1, 16, 16, 1], [1, 8, 8, 1], [1] * 4)` passed in to the Op `tf.image.extract_patches `. Here the shape only inferrable after calling with some input data which do some computation.This computation won't happen during Jit compilation. Hence you may get error.

If you can build the model such a way that the output shapes independent of actual computation then jit_compile should work.Please review once again and comeback if anything missing or needs clarification here.

Thankyou!

**Comment by shkarupa-alex** [comment]
@SuryanarayanaY i figured out that issue comes from gradient computation.
Here is a fix https://github.com/tensorflow/tensorflow/pull/59185

**Comment by SuryanarayanaY** [comment]
I have validated my [comment](https://github.com/tensorflow/tensorflow/issues/59058#issuecomment-1383724816) above.I tried to pass the Inputs with clear input shape `(16, 16, 8)` and model.summary() prints the output shape as `(None, 2, 2, 1)` which means all the shapes are inferrable at the compilation time itself.Please refer attached revised [gist](https://colab.research.google.com/drive/1dV514BI2v04SSiH_f37zlPjfQldcuOlN?resourcekey=0-Cc5MLY8wDXlFZkGr1eOw1Q#scrollTo=_j9EkheY0Csd).

```
Model: "model_4"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
=================================================================
 input_5 (InputLayer)        [(None, 16, 16, 8)]       0         
                                                                 
 temp_4 (Temp)               (None, 2, 2, 1)           8481      
                                                                 
=================================================================
Total params: 8,481
Trainable params: 8,481
Non-trainable params: 0
```

Hence this problem is not related to shape inference as suspected earlier by me.May be this is valid bug which needs fix.Hoping the above commit by @shkarupa-alex fixes this issue.Thanks!
